### PR TITLE
Add atomic definitions

### PR DESCRIPTION
**Pull request by dcodeIO**
fixes https://github.com/AssemblyScript/assemblyscript/issues/523

### CODE REVIEW

**Comment by MaxGraey** [comment]
What do you think about add special entity (decorator) for enabling wasm feature from code as well?
Like:
```ts
features.enable("threads simd")
// or other variant
// ASC_FEATURES("threads simd")

// use some atomic & SIMD operations
```

**Comment by dcodeIO** [comment]
Not sure. One usually compiles for a modern browser or a very specific target. In modern browsers, all these features will be available by default eventually anyway, and everything else seems sensible to bury behind a compiler switch (becoming `--disable` then). Also, doing this in code could become quite complex once multiple dependencies are in play and I'd perfer to have more control over this, like provided by a compiler flag / multiple compile steps with different arguments.

**Comment by dcodeIO** [comment]
Some notes:

* Inline-assembler-like functions now use native types exclusively
* One thing that's still odd is `i64.atomic.rmw8.cmpxchg_u(0, <i64>1, 2);` emitting an `i32.` instruction if the argument is a literal not explicitly made `<i64>` (becomes i32 then)
* There are new helpers to get diagnostic ranges of typeArguments and arguments that we should use in other places as well
* `wait` and `notify` don't validate in nightly yet

File: src/ast.ts [hunk 1]
```diff
@@ -1350,6 +1350,24 @@ export class CallExpression extends Expression {
...
+      return Range.join(typeArguments[0].range, typeArguments[typeArguments.length - 1].range);
+    }
+    return this.expression.range;
+  }
+
+  /** Gets the arguments range for reporting. */
+  get argumentsRange(): Range {
+    var args = this.arguments;
+    if (args.length) {
+      return Range.join(args[0].range, args[args.length - 1].range);
```
Comment (MaxGraey) [inline src/ast.ts]: may be make sense cache `var len = args.length`?

Comment (dcodeIO) [reply src/ast.ts]: sure, can do that. have not been optimizing diagnostics so much because it only displays in error cases anyway :)

Comment (MaxGraey) [reply src/ast.ts]: yeah, but when AS will be compile by itself it probably could reduce size a little)

File: src/builtins.ts [hunk 2]
```diff
@@ -2188,48 +2191,49 @@ export function compileCall(
...
+        loadType.byteSize,
         arg0,
-        typeArguments[0].is(TypeFlags.INTEGER) &&
+        loadType.is(TypeFlags.INTEGER) &&
         contextualType.is(TypeFlags.INTEGER) &&
-        contextualType.size > typeArguments[0].size
+        contextualType.size > loadType.size
           ? (compiler.currentType = contextualType).toNativeType()
-          : (compiler.currentType = typeArguments[0]).toNativeType(),
+          : (compiler.currentType = loadType).toNativeType(),
```
Comment (MaxGraey) [inline src/builtins.ts]: Hmm, it's little bit unusual style. May be better:
```ts
compiler.currentType = contextualType.size > loadType.size 
  ? contextualType.toNativeType()
  : loadType.toNativeType()
```

Comment (dcodeIO) [reply src/builtins.ts]: Hmm, that would assign a `NativeType` (enum/i32) to a `Type` (object).

Comment (MaxGraey) [reply src/builtins.ts]: Ahh, miss this a complex expression)

**Pull request merged**